flink-connector-vitess-cdc/src/test/java/com/vervetica/cdc/connectors/vitess/VitessSourceTest.java

```diff
@@ -0,0 +1,310 @@
+            throws Exception {
+        List<T> allRecords = new ArrayList<>();
+        LinkedBlockingQueue<StreamRecord<T>> queue = sourceContext.getCollectedOutputs();
+        while (allRecords.size() < expectedRecordCount) {
+            StreamRecord<T> record = queue.poll(1000, TimeUnit.SECONDS);
+            if (record != null) {
+                allRecords.add(record.getValue());
+            } else {
+                throw new RuntimeException(
+                        "Can't receive " + expectedRecordCount + " elements before timeout.");
+            }
+        }
+
+        return allRecords;
+    }
+
+    private boolean waitForSourceToStart(
+            Duration timeout, DebeziumSourceFunction<SourceRecord> source)
+            throws InterruptedException {
+        long now = System.currentTimeMillis();
+        long stop = now + timeout.toMillis();
+        while (System.currentTimeMillis() < stop) {
+            if (source.getDebeziumStarted()) {
+                break;
+            }
+            Thread.sleep(10); // save CPU
+        }
+        Thread.sleep(10000); // Wait for full start
+        return source.getDebeziumStarted();
+    }
+
+    private static <T> void setupSource(DebeziumSourceFunction<T> source) throws Exception {
+        setupSource(
+                source, false, null, null,
+                true, // enable checkpointing; auto commit should be ignored
+                0, 1);
+    }
+
+    private static <T, S1, S2> void setupSource(
+            DebeziumSourceFunction<T> source,
+            boolean isRestored,
+            ListState<S1> restoredOffsetState,
+            ListState<S2> restoredHistoryState,
+            boolean isCheckpointingEnabled,
+            int subtaskIndex,
+            int totalNumSubtasks)
+            throws Exception {
+
+        // run setup procedure in operator life cycle
+        source.setRuntimeContext(
+                new MockStreamingRuntimeContext(
+                        isCheckpointingEnabled, totalNumSubtasks, subtaskIndex));
+        source.initializeState(
+                new MockFunctionInitializationContext(
+                        isRestored,
+                        new MockOperatorStateStore(restoredOffsetState, restoredHistoryState)));
+        source.open(new Configuration());
+    }
+
+    private static class ForwardDeserializeSchema
+            implements DebeziumDeserializationSchema<SourceRecord> {
+
+        private static final long serialVersionUID = 2975058057832211228L;
+
+        @Override
+        public void deserialize(SourceRecord record, Collector<SourceRecord> out) throws Exception {
+            out.collect(record);
+        }
+
+        @Override
+        public TypeInformation<SourceRecord> getProducedType() {
+            return TypeInformation.of(SourceRecord.class);
+        }
+    }
+
+    private static class MockOperatorStateStore implements OperatorStateStore {
```

Comment:
nit: Could you move these code in the test-util package?

Reply:
it would require refactoring of other modules not only vitess, I would prefer to do this in a separate pull request